pandas/core/indexing.py

```diff
@@ -2310,6 +2310,51 @@ def check_bool_indexer(ax, key):
     return result
 
 
+def check_setitem_lengths(indexer, value, values):
+    """Validate that value and indexer are the same length.
+
+    An special-case is allowed for when the indexer is a boolean array
+    and the number of true values equals the length of ``value``. In
+    this case, no exception is raised.
+
+    Parameters
+    ----------
+    indexer : sequence
+        The key for the setitem
+    value : array-like
+        The value for the setitem
+    values : array-like
+        The values being set into
+
+    Returns
+    -------
+    None
+
+    Raises
+    ------
+    ValueError
+        When the indexer is an ndarray or list and the lengths don't
+        match.
+    """
+    from pandas.core.indexing import length_of_indexer
+
```

Comment:
you need to assert ``com.is_bool_indexer`` here (and you can actually remove some of this logic as it duplicates).

further should actually move ``is_bool_indexer`` to ``pandas.core.indexing``

Reply:
I think `com.is_bool_indexer` serves a different purpose, and should already have been called typically when ending up in this place. 
This helper function is only for generating informative error messages when the length does not match, `com.is_bool_indexer` does actual conversion and inference of the indexer.

Reply:
of course, but its serving a very similar purpose to this and needs to be moved here.

Reply:
`is_bool_indexer` seems to accept list-like, while this explicitly needs an ndarray since it checks `indexer[indexer]`. Given that I need to do `isinstance(indexer, ndarray)` anyway, to ensure that that boolean masking works, to tradeoff is

```python
            if not (isinstance(indexer, np.ndarray) and
                    indexer.dtype == np.bool_ and
                    len(indexer[indexer]) == len(value)):
 
```

vs.

```python
            if not (isinstance(indexer, np.ndarray) and
                    com.is_bool_indexer(indexer) and
                    len(indexer[indexer]) == len(value)):
 ```

Since the first is likely to be faster, I'd prefer to just go with that.